src/Compilers/CSharp/Portable/Binder/DecisionDagBuilder.CheckOrReachability.cs

```diff
@@ -0,0 +1,956 @@
+                        patternIndex++;
+                    }
+                }
+            }
+
+            existingCases.Free();
+        }
+
+        private static void CheckOrAndAndReachability(
+          ArrayBuilder<StateForCase> previousCases,
+          int patternIndex,
+          BoundPattern pattern,
+          DecisionDagBuilder builder,
+          BoundDagTemp rootIdentifier,
+          LabelSymbol defaultLabel,
+          SyntaxNode syntax,
+          BindingDiagnosticBag diagnostics,
+          bool isSwitchExpression = false)
+        {
+            CheckOrReachability(previousCases, patternIndex, pattern,
+                builder, rootIdentifier, defaultLabel, syntax, diagnostics, ignoreDefaultLabel: isSwitchExpression);
+
+            CheckOrReachability(previousCases, patternIndex, MoveNotPatternsDownRewriter.MakeNegatedPattern(pattern),
+                builder, rootIdentifier, defaultLabel, syntax, diagnostics, ignoreDefaultLabel: isSwitchExpression);
+        }
+
+        private static void CheckOrReachability(
+            ArrayBuilder<StateForCase> previousCases,
+            int patternIndex,
+            BoundPattern pattern,
+            DecisionDagBuilder builder,
+            BoundDagTemp rootIdentifier,
+            LabelSymbol defaultLabel,
+            SyntaxNode syntax,
+            BindingDiagnosticBag diagnostics,
+            bool ignoreDefaultLabel = false)
+        {
+            var normalizedPattern = MoveNotPatternsDownRewriter.Rewrite(pattern);
+
+            SetsOfOrCases setOfOrCases = RewriteToSetsOfOrCases(normalizedPattern);
+            if (setOfOrCases.IsDefault)
+            {
+                return;
+            }
+
+            // We construct a DAG and analyze reachability of branches once per `or` sequence
+            Debug.Assert(!setOfOrCases.IsDefault);
+            foreach (OrCases orCases in setOfOrCases.Set)
+            {
+                using var casesBuilder = TemporaryArray<StateForCase>.GetInstance(orCases.Cases.Count);
+                var labelsToIgnore = PooledHashSet<LabelSymbol>.GetInstance();
+                PopulateStateForCases(builder, rootIdentifier, previousCases, patternIndex, orCases, labelsToIgnore, syntax, ref casesBuilder.AsRef());
+                BoundDecisionDag dag = builder.MakeBoundDecisionDag(syntax, ref casesBuilder.AsRef());
+
+                foreach (StateForCase @case in casesBuilder)
+                {
+                    if (!dag.ReachableLabels.Contains(@case.CaseLabel) && !labelsToIgnore.Contains(@case.CaseLabel))
+                    {
+                        diagnostics.Add(ErrorCode.WRN_RedundantPattern, @case.Syntax);
+                    }
+                }
+
+                if (!ignoreDefaultLabel && !dag.ReachableLabels.Contains(defaultLabel))
+                {
+                    diagnostics.Add(ErrorCode.WRN_RedundantPattern, syntax);
```

Comment:
>diagnostics.Add(ErrorCode.WRN_RedundantPattern, syntax); [](http://example.com/codeflow?start=20&length=56)

It looks suspicious that we are doing this check for every set of cases. It feels like, at the very least, this might report duplicate errors when we have more than one set. Also, it is not clear why this check is performed only when we have sets. If the pattern is always true (I assume this is what we are after here), I think reporting the fact shouldn't really depend on whether the pattern produces any sets of cases